mesh/mesh_connections.hpp

```diff
@@ -0,0 +1,95 @@
+class MeshConnections
+{
+   private:
+   Mesh *mesh;
+   Table *elem_to_elem, *elem_to_face, *elem_to_edge, *elem_to vert;
+   Table *face_to_elem, *face_to_face, *face_to_edge, *face_to vert;
+   Table *edge_to_elem, *edge_to_face, *edge_to_edge, *edge_to vert;
+   Table *vert_to_elem, *vert_to_face, *vert_to_edge, *vert_to vert;
+
+   public:
+   //Queries
+   bool IsFaceInElem(int elem, int face);
+   bool IsEdgeInElem(int elem, int edge);
+   bool IsVertInElem(int elem, int vert);
+   bool IsEdgeInFace(int face, int edge);
+   bool IsVertInFace(int face, int vert);
+   bool IsVertInEdge(int edge, int vert);
+
+   //Neighbor mappings
+   void NDNeighbors(int nd, int id, Array<int> &neighbors);
+   void NeighborsToElem(int elem, Array<int> &elems);
+   void NeighborsToFace(int face, Array<int> &faces);
+   void NeighborsToEdge(int edge, Array<int> &edges);
+   void NeighborsToVert(int vert, Array<int> &verts);
```

Comment:
What is meant by neighbors? Are two elements neighbors if they share a face? Edge? Vertex?

The more general version of this is a query like: give me all of the `dim`-dimensional entities that share a `sdim`-dimensional entity with the given entity.

Reply:
Yeah what I am driving at with these is entities that share a dim-1 with the given entity which is typically what folks think of as neighbors (at least element neighbors).  Although your generalization would be valuable
NDNeighborsSharingMDParts?

Reply:
I'm definitively open to suggestions on the names of these things.  Parts->Entities?

Reply:
My suggestion would be to rename `NDPartsInMDPart` simply `Children` and `NDPartsWithMDPart` to `Parents`.

A mesh entity is uniquely determined by the pair `(dim, id)` (i.e., in a 3D mesh, the pair `(3, i)` would identify element `i`, `(2, i`) would identify face `i`, and so on). We could either have an entity be `std::pair<int,int>` or just a simple struct.

Then the signature would be something like this (and the "multiple" queries could just be overloads):

```c++
/// Return all indices of dim-dimensional children of the given entity
void Children(Entity entity, int dim, Array<int> &ids);
void Children(const Array<Entity> &entities, int dim, Array<int> &ids);

/// Return all indices of dim-dimensional parents of the given entity
void Parents(Entity entity, int dim, Array<int> &ids);
void Parents(const Array<Entity> &entities, int dim, Array<int> &ids);
```

So to get the vertices of the `i`th element, you would just call
```c++
Children({dim, i}, 0, vertex_list);
```
and to get the list of all elements containing edge `j`, you would call
```c++
Parents({1, j}, dim, element_list);
```

For the neighbors, the signature could look like
```c++
/// Return the indices of all dim-dimensional neighbors of the given
/// entity, sharing a sharing_dim dimensional interface
Neighbors(Entity entity, int dim, int sharing_dim, Array<int> &ids);
```

This way, the face, edge, and vertex neighbors of an element would be obtained by
```c++
Neighbors({dim, i}, dim, dim-1, face_neighbors);
Neighbors({dim, i}, dim, 1, edge_neighbors);
Neighbors({dim, i}, dim, 0, vertex_neighbors);
```
It also enables more general queries, like give me all edges sharing a vertex with the given element
```c++
Neighbors({dim, i}, 1, 0, edge_list);
```

As an aside, another reason why I tend to prefer an approach like the above is that it avoids the ambiguity around the meaning of "faces" that is already present in MFEM: `GetNFaces` returns the number of 2D entities in a 3D mesh, and `GetNumFaces returns the number of `dim-1` entities in a `dim` mesh. When simply referring to entities by the integer dimension, there is no such ambiguity.

Reply:
I see what you mean here.  How would you handle the covered by and touched by situations?
ParentsCoveredByChildren
ParentsTouchedByChildren
??

Reply:
And as far as the entities go what about boundary entities.

Reply:
> I see what you mean here. How would you handle the covered by and touched by situations?
> ParentsCoveredByChildren
> ParentsTouchedByChildren
> ??

Right, those are somewhat different queries, it's something like `ParentsOfAny`, which would take a list of entities, and return a list of entities that contain *any* of the given ones, and `ParentsOfAll`, which would take a list of entities, and return a list of entities that contain *all* of the given ones.

Reply:
In terms of boundary entities, in MFEM only codimension-1 entities can be marked as boundary elements, which might be a bit of a restriction. In that sense, in a 3D mesh, MFEM doesn't really have the concept of a boundary edge or vertex.

Reply:
The boundary entities can get a flag I guess {dim, i, bdrflag}.  Then we could try breaking the query into 3 pieces the kind of entity you want, the kind of entity you have, and the relationship you want to expose.